src/basicfuns.jl

```diff
@@ -44,6 +44,36 @@ function xlog1py(x::Number, y::Number)
     return iszero(x) && !isnan(y) ? zero(result) : result
 end
 
+"""
+$(SIGNATURES)
+
+Return `x * exp(x)` for `x > -Inf`, or zero if `x == -Inf`.
+
+```jldoctest
+julia> xexpx(-Inf)
+0.0
+```
+"""
+function xexpx(x::Real)
+    expx = exp(x)
+    return iszero(expx) ? expx : x * expx
+end
+
+"""
+$(SIGNATURES)
+
+Return `x * exp(y)` for `y > -Inf`, or zero if `y == -Inf`.
+
+```jldoctest
+julia> xexpy(1.0, -Inf)
+0.0
+```
+"""
+function xexpy(x::Real, y::Real)
+    expy = exp(y)
+    return iszero(expy) && !isnan(x) ? zero(x * expy) : x * expy
```

Comment:
What's the reason for changing it to `zero(x * expy)`? To avoid the promotion? Similar to `xlogy` etc. maybe one could even use
```suggestion
    result = x * expy
    return iszero(expy) && !isnan(x) ? zero(result) : result
```
Possibly this is faster if branch prediction fails.

Reply:
I figure `zero(x * expy)` was more readable. Since ternary operator short-circuits the product `x * expy` will not be evaluated more than once anyway.

Reply:
I'm pretty sure that the ternary operator does not short-circuit, and only logical operators such as `&&` and `||` do. In contrast to `ifelse` both branches don't have to be evaluated (there per se it's a regular function call and hence arguments have to be evaluated), but there's no guarantee that the branch prediction of the CPU always guesses the correct branch and only one of the two branches is evaluated.

Reply:
Oh you are right, ternary operator does not short-circuit. But I'd guess common sub-expressions would be extracted from the branch by the compiler. Anyway the `result = ...` trick is used also in `xlogx`, etc., so I went with that.